### DISCUSSION TITLE
Suggestion: commercialize FeathersJs

### DISCUSSION THREAD

**Discussion by haniyasin** [feathersjs/feathers#2353] [Ideas]
it seems the contributors are busy building , i would like to suggest commercializing Feathersjs for better adoption by organizations in a form of a yearly subscription like Telerik, Syncfusion, Devexpres,... the yearly subscription is affordable to most Startups, and small businesses , yet it offers a great gate for support where all Questions and support requests get answered 
asking for help in an hourly bases for USD 100-200 /hour is affordable to bigger enterprises i am not one of those
the sponsorship scheme outcome of 1 -2 hours per month of support is not worth it

**Comment by tvld** [comment]
The reason why we chose Feathersjs is because of its lightweight, amazing elegance ...and awesome management of the founder. He is enormously responsive and knowledgeable. Of the 100 or so open-source frameworks and libs I used in my life, he stands out. If possible, I would be grateful if this could stay the same...

Splitting the code in open-source/commercial is recipe for disaster. And paid support = non-support if not paid for leads to nowhere in my opinion. Too quickly questions remain dangling because they should have been asked to pay, and so on...

If commercialization needs to be, and continuity might require it one day, I rather see it in the form of some commercial hosting /SAAS of added value that is _outside the scope_ of the core _as well as_ the add-ons. For example, a recent discussion about 2FA could develop into commercial service. Optionally some MongoDB backend hosting could be beneficial, etc.. If done properly by the founders, this would surely be taken on and I see an enormous bright future for Feathers))).

**Reply by haniyasin** [reply]
i did not mean a closed source sort of commercialization , only the support part , if you have noticed recently feathers has a discussion part where non bugs are re-routed there , the Qs can range from novice to master in complexity, the contributors might rely on the community to answer most of them , it is a fact that not all get answered perfectly  or not at all
i respect feathers architecture of not adopting the MVC style it is elegant and very productive, robust , I tried it in a POC for intercepting IoT zigbee sensors in a factory, data is collected locally 3 data values recurring each 15 seconds from 100 sensor/router  divided between 2 feathers services that is 1200 point/ minute connected through sockets to 2 vuejs dashbords with real time event update 
but i wounder if i am not an official developer ,and  if i hire/not hire developers for developing an LoB system targeting the enterprise , and my team or i need support (we can not wait days or weeks to get an answer , i can not afford usd 200 /hr , and the maximum contribution will not give me enough , what should i do (thinking loud , should i better subscribe to a commercially supported framework like XAF even if it is in another tech stack , or CUBA platform or maybe nestjs  , or should i stick with what i admire and respect , but again wait i might get stuck and am not rich )
that is the story behind my suggestion :)

**Reply by daffl** [reply]
Ok, I'm going to add my quick thoughts on sustainability here after all 😄 Here is the thing that has become fairly clear (but non-open-source developers might not be aware of as much): Open Source as it is right now is not sustainable. There is of course much more to say about this but you just have to look around to see that the space is basically being monopolized by:

- Mega tech companies (Facebook React/GraphQL, Microsoft TypeScript, Google Kubernetes/Angular, AWS)
- Open source by VC funded companies (Next, Gatsby, Apollo, Hasura, Strapi etc. basically many of the things you hear about all the time - gotta put the marketing 💰 💵 somewhere 😆). Some of those are great projects but those VCs want to see their returns somehow at some time.

@tvld I agree hosting would be a good option. Interesting that you mention using MongoDB because it is one of the open source projects that _did_ in fact commercialize their license with their standard license not allowing other hosting services to use it. They were (rightfully so) afraid that Amazon just comes around, turns it into a service and crushes them. My questions here are: When isn't the question of having to pay dangling around? There is always some kind of cost to using open source, no? And why is the question of paying for things such a bad thing?

**Reply by tvld** [reply]
In general, I think that "paying for things" is in principle a _good_ thing. It can increase the chance of continuity and level of QoS. It also would provide a well-deserved reward for the core staff who invested so many free hours/years...
In many cases, I have seen open source fade away by improper execution of commercialization. Marketing happens to be my main [interest/expertise](https://www.bligson.com) and with that background, I would suppose that it can and maybe should be done, assuming:

- open source communities tend to consist of people that do appreciate trying out things and sharing thoughts for free. These people should not be alienated. They purposely chose a non-commercial environment in the first place. Losing the early adopters will cost the project dearly and will destroy the "lightness of being"
- as you say, legal boundaries should prevent the big boys, and surely at least the 'evil' big five, to hijack the commercialization. Indeed without any doubt, they _will_ come around as soon as success reaching a tipping point
- the Feathers framework in particular takes an interesting place in the stack: close to infrastructure. For this, I would not be surprised if 90%+ of all projects are from people fiddling in their garage, basement, or even attic. They don't care so much about up-time or security.
- I would not be surprised if the vast majority of the Feathers users can most likely not afford to pay...until their own project becomes a "hit" in one way or another. The chances are slim like with any start-up project but if so, they will very _gladly_ pay as their existence hinges on it.

And there we might have found the best opportunity for scalable money-making. Because suddenly the database should not be offline anymore, 2FA is required, load balancing, and so on. Basically, the moment any project is a hit, the backend framework becomes the heart, lifeblood and the whole company might depend on it to work reliable, fast, and safe. And there is the money !

**Comment by daffl** [comment]
This is a very relevant discussion which I also have some more news to share with soon. I also was about to write a long post about open source sustainability (or lack thereof) but then decided that it is not really relevant here.

As for the rate, if you are saying 150 - 200 USD/hr is too much then I don't think there is much more discuss. It's a pretty standard consulting rate for North America and given that everybody is already using thousands of hours of my work and experience for free probably on the low end. I can definitely say that people I helped in an hour said it saved them at least a day or more of their own time. I am also quite certain you won't get real quality support for less with any the other platforms you mentioned.

**Reply by haniyasin** [reply]
am sorry for sticking my nose into , may be i should not, but i do care,  am not in your shoes, nor i know the challenges you might face on a daily basis, yet i bet they are common world wide "making a living to live in dignity" 
if you could harness for example 200 subscriptions as a start for 1500 - 2000 USD/yr , that is 300K to 400K/yr , if you share with other great contributors within feathers ecosystem like Marshall from feathers-vuex and others you will stand as a united team, this will lower the pressure on you as a team to seek a living, free up more time to development

the SLA would target (naturally ) Smaller and non-VCed startups, to a certain complexity point , and am pretty sure guys like you (the team) would to take minutes to solve most of the requests while you are having your morning espresso in a nice café.

you can still charge whatever  consulting fee relevant to NA for deeper consultations ( or other higher end customers) for issues like architecture, clustering, distributing, ....
you can use part of it for marketing to convert more subscribers , then delegate support to a hired team with an escalation plan.

i had once and still subscription with some commercially supported platforms/libraries, it is true that sometimes i had to hit my head against a wall , but for the most part the support team was always keen to help and share working code samples even for the most trivial cases.

in all cases i wish you the best.

**Reply by daffl** [reply]
That subscription level would be totally reasonable and what the existing 100$/mo tier at https://github.com/sponsors/daffl was aiming for. The way things are right now, _anybody_ who is a supporter at any level gets priority with their issues and questions. I'm also planning to adjust those tiers and add one-time options (which GitHub only added only recently). This is different than an hourly commitment though so I think that's where there was a bit of a misunderstanding.

**Reply by haniyasin** [reply]
I salute you, hope this really work for the best of feathers and adopters

**Reply by tvld** [reply]
Payment per hour, how high or low the fee might be is IMHO _not_ a good idea for the core team.  There are two reasons why not.

1. there hardly is a "fair price": We can argue that even $1000 per hour is not enough considering all work done and the great inventions. So, given the high but fair price, _who_ will pay then, who can afford it? 
2. scalability, see my point below

_By the way, per hour outsourcing is great for the second-order supporters I think. These are the ones that scale the project for a living._

**Reply by tvld** [reply]
@daffl I have great admiration for the high level of responsiveness you give on all the forums and so on. It makes you stand out. By commercialization of support questions, I fear you might lose this openness, free-thinking, and direct community interaction, but I migth be wrong.

Are you sure you really want to deviate from the strategic development dreams and freedom of choice that brought you here in the first place? You might risk trading your freedom for triviality...

**Reply by tvld** [reply]
...   2. All in all, from a strategic point of view as SAAS is infinitely scalable, I think the core team should get their reward and return of investment by scale. And let's face it, scalability is _not_ supported in tiers of service per hour.

We know that Feathers users _will_ need professionalization as soon as they hit some kind of homerun, on their own or for a big-pocket client. And there lies the opportunity, I think

So, my suggestion, for what it is worth: keep all the same as is. Keep your freedom, throw in your support here and there. Learn, play... But then additionally offer a _strong and scalable_ commercial basis for all the viable projects that blossom from the community. 

Maybe we could think of services such as:

1. true scalability, load-balancing
2. monitoring
3. backup
4. security (2FA)
5. integration with _commercial_ third-party services
6. ...

I do not mention simple database hosting and stuff. I would not offer things in paid plan that simple starters will do themselves for free, such as with a simple Docker run. Let others do that 'boring' stuff. You stand out so make sure also the new people will recognize you as such...;

_The answer is forward, not backwards... well these are my two cents in this discussion... ;-)))_

**Reply by J3m5** [reply]
@tvld I'm not sure if going on the Saas route is really what users need, at least the big ones.
Considering the two trends we are witnessing there days, which are k8s and p2p, I don't see how a SaaS offer would position itself.
There are already plenty of services that offers nodejs hosting with the features you mentioned and with a k8s architecture those points are handled at the cluster level.
Building a SaaS only around feathers would be quiet limiting IMO.

**Reply by daffl** [reply]
This is a good discussion, I appreciate everybody's thoughts.

I think both avenues have advantages and disadvantages. I've been doing paid support and consulting for a while now and would like to try and grow it a bit more. It is something companies often explicitly ask for, because things often can't be shared in a public forum, to have a more direct line of communication or to get training specifically for their teams and use cases. It is also a great way to get insights into common problems and makes for a positive feedback loop where many improvements have been included that way in the open source stack over the years.

SaaS is interesting because it would incentivize working on things that would just be too complex to handle as an open source project. Authentication is a great example, where the sale of Auth0 for 6.5 billion (!) dollars gives rough idea of the size and complexity of the problem space. I also don't believe that K8s is the end of technology and a solution for everything and even less so what's being sold to us as "serverless" which is basically just complex proprietary (hosted) technology stacks. All of those are to more than 95% dominated by three platforms - AWS, Google Cloud and Azure which also happen to be owned by the three most valuable companies in the world. Whenever there is such a massive monopolization there is always an opportunity to do something different and maybe even better ("The Master Switch" by Tim Wu an interesting read related to that). I think instead of throwing the towel because you can never compete, it's worth looking at their blind spots. Personally where I see opportunity here would be:

- Picking up where Heroku left off. Heroku is still seen as kind of a "Holy Grail" in ease-of-use but [definitely dropped the ball on innovation after being sold to SalesForce](https://www.infoworld.com/article/3614210/the-decline-of-heroku.html?page=2). After having to do it a couple of times, I honestly don't believe that anybody _actually_ likes digging through the AWS or Google Cloud consoles and poking at K8s configuration files until things maybe work (and then basically creating their own internal Heroku clone). To me it always felt like a solution to a problem that shouldn't exist in the first place.
- Growing P2P technologies and helping to scale and bridge to the "old web". There may be a bit of a buzz around this (I'm all in for it and definitely believe that it will change the way we build things in the next few years) but I can't stress how _early_ it is. Most projects are prototypes that get better every day but are still far from production ready. I do however think it is possible to leverage some those technologies already to realize some of its potential. Examples would be using IPFS for secure, predictable and almost instant deployments (more about that maybe soon 😄) or an authentication system that uses the web-crypto API and public/private keys under the hood but allows to link to traditional user/oAuth accounts. Eventually I'd like to see an infrastructure that allows _any_ developer in the world to start putting their own app online by just sharing a link (this is possible!) and without having to get a masters degree in AWS console or requiring a credit card to sign up.

Either way, this will definitely be an interesting year because some ideas are cooking 🍵 Worst case, none of those things work out and it's time for me to move on to opening a pretzel bakery and beer garden on one of the gulf islands here on the west coast 🥨 🍻

**Reply by haniyasin** [reply]
> @daffl I have great admiration for the high level of responsiveness you give on all the forums and so on. It makes you stand out. By commercialization of support questions, I fear you might lose this openness, free-thinking, and direct community interaction, but I migth be wrong.
> 
> Are you sure you really want to deviate from the strategic development dreams and freedom of choice that brought you here in the first place? You might risk trading your freedom for triviality...

My Friend, you can never have a free thinking if your mind is busy trying to secure your family where ever you might settle OECD or not , life is tough and it is getting tougher and tougher, unless some one get Shell to dig an oil well in his backyard with a guaranteed paycheck every month

**Comment by haniyasin** [comment]
i would like to share samples of some commercially supported platforms to help grasp a feeling of what is happening in this market from other tech stacks
https://serenity.is/#Pricing 
this is their pricing, the support is not limited to the number of incidents mentioned it mentions the guaranteed response time for this quantity
this is their SLA https://serenity.is/docs/startsharp/README 

there is also https://abp.io/ open source with commercial license https://commercial.abp.io/pricing , not a good SLA in my opinion and they are pricier but it can give a good feel of what the commercially supported platforms are doing
not to forget XAF from devexpress around 2xxx/yr but give you access to all their components stack 

i saw a node based framework with this type of support scheme but i forgot which one it was

this is cuba platform https://www.haulmont.com/services/cuba-platform-services/support/ 
pricier seems they are not interested in less fortunate startups but again can give a feeling

i may also urge to construct an official training or boarding service on-demand /1-to-1 , i saw coding garden but they were introductory oriented

a well thought startup whether bootstrapped or financed know for sure that you have to spend money to make money, unless they are super JS gurus who can stand up on their own. 
for the rest of us we need somebody to lean on :)
